### DISCUSSION TITLE
ch9/ch9-04

### DISCUSSION THREAD

**Discussion by giscus** [gopl-zh/gopl-zh.github.com#186] [General]
# ch9/ch9-04

<The Go Programming Language>中文版

https://golang-china.github.io/gopl-zh/ch9/ch9-04.html

**Comment by ccsuper1024** [comment]
这里后面写的这么复杂，反而不知道在说啥。其实就是因为多核cpu中，并发运行时，在编译器编译后x,y变量有可能是在两个独立的CPU上都有副本的，并且此时是被初始化为0的，再之后，由于编译器认为A1,A2这两条语句和B1,B2这两条语句的顺序不影响结果，就有可能调换两者的次序。从而打印出x:0 y:0，y:0 x:0

**Comment by lazybearlee** [comment]
这里可以去看看golang的内存模型
[Golang内存模型](https://go.dev/ref/mem)

**Comment by coderZoe** [comment]
听着有些像Java的可见性与`volatile`

**Comment by Etherealss** [comment]
也就是说，即使在 goroutine A 中**已经**执行了 A1,A2，但 goroutine B 未必能感知到 A1 已经执行了，它可能还是会输出它所在的 CPU 核心缓存中的值：`x=0`

**Comment by Zacharia2** [comment]
并发编程的核心难点不仅是操作的执行顺序，更是内存的可见性。现代CPU的缓存和编译器的优化，可能导致一个goroutine对变量的修改，无法被其他goroutine及时、按序地看到，即使代码逻辑上看起来没问题。因此，即便是一个简单的只读操作，也必须通过互斥锁或channel等同步原语来强制CPU刷新缓存、保证内存视图的一致性，从而避免数据竞争和不可预测的结果。
